File: sort/merge_sort.cpp
#include<bits/stdc++.h>

class Solution {
public:
    std::vector<int> mergeSort(std::vector<int>& arr) {
        int lo=0, hi=arr.size()-1;
        sort(arr, lo, hi);
        return arr;
    }
    void sort(std::vector<int>& arr, int lo, int hi) {
        if (lo >= hi) return;
        int mid = lo + (hi-lo)/2;
        sort(arr, lo, mid);
        sort(arr, mid+1, hi);
        merge(arr, lo, mid, hi);
    }

    void merge(std::vector<int>& arr, int lo, int mid, int hi) {
        int i=lo, j=mid+1;
        std::vector<int> aux(hi-lo+1);
        for (int k=lo; k <=hi; k++) {
            aux[k] = arr[k];
        }
        for (int k=lo; k<=hi; k++) {
            if (i > mid) 
                arr[k] = aux[j++];
            else if (j > hi) 
                arr[k] = aux[i++];
            else if (aux[i] < aux[j])
                arr[k] = aux[i++];
            else  
                arr[k] = aux[j++];
        }
    }

};

int main() {
    std::vector<int> arr = {1, 4, 10, 3, 6, 8, 9, 2, 5};
    Solution so;
    std::vector<int> result = so.mergeSort(arr);
    for (int i=0; i < result.size(); i++) {
        std::cout << result[i] << " ";
    }
    std::cout << std::endl;
}

File: sort/merge_sort.py
class Solution:
    def divide(self, arr, lo, hi):
        if lo==hi: return [arr[lo]]
        mid = lo + (hi-lo)//2
        nums1 = self.divide(arr, lo, mid)
        nums2 = self.divide(arr, mid+1, hi)
        return self.merge(nums1, nums2)
    
    def merge(self, nums1, nums2):
        nums = []
        i, j = 0, 0
        m, n = len(nums1), len(nums2)
        while i < m and j < n:
            if nums1[i] < nums2[j]:
                nums.append(nums1[i])
                i+=1
            else:
                nums.append(nums2[j])
                j+=1
        if i < m:
            nums += nums1[i:]
        if j < n:
            nums += nums2[j:]
        return nums


if __name__ == "__main__":
    so = Solution()
    arr = [1, 4, 3, 2, 5]
    lo, hi = 0, len(arr)-1
    print(so.divide(arr, lo, hi))

